### ISSUE TITLE
Unauthorized when opening the IDE

### ISSUE THREAD

**Issue by l0rd** [eclipse/che#22353]
### Describe the bug

![2023-07-11 12 14 57](https://github.com/eclipse/che/assets/606959/02d144d7-b015-4c8c-9abb-0fcc1e7952bd)

This happened on sandbox this morning

### Che version

7.70@latest

### Steps to reproduce

Open https://github.com/eclipse-che-demo-app/che-demo-app on sandbox

### Expected behavior

No error

### Runtime

OpenShift

### Installation method

OperatorHub

### Environment

Linux

**Comment by ibuziuk** [comment]
similar 404 error that was caught by periodic tests - [output.webm](https://github.com/eclipse/che/assets/1461122/e565147b-1850-4b74-b1d4-4ce76e947222)

**Comment by dkwon17** [comment]
I see similar issues (also similar to https://github.com/eclipse/che/issues/22352) when refreshing the dashboard
<img width="1214" alt="image" src="https://github.com/eclipse/che/assets/83611742/50f63ee6-953b-4f68-a8a8-6d6eeecaea15">

This error is quite difficult to reproduce.  On some days, I'm able to reproduce this issue very occasionally on the dogfooding cluster, but on other days, I cannot reproduce it at all (I've used a script to refresh the dashboard on dogfooding about 4000 times, and could not reproduce it). I find it much easier to reproduce this error on Dev Sandbox, more specifically the m2 cluster.

I've noticed that these types of errors often occur because some requests made by the dashboard occasionally returns a 401 or 403 error.

<details><summary>403 error</summary>

Sometimes the request to get the user namespace results in a 403 error:

![image](https://github.com/eclipse/che/assets/83611742/d5fe16d2-9ad6-49fe-8b91-5b39dbc18519)

![image](https://github.com/eclipse/che/assets/83611742/35105710-22cd-4b60-9b56-e39dcbb9cf26)

This causes other requests like `events` and `pods` to fail as well, because they both depend on the value of the user namespace. 

In this response header, there are some differences compared to when the `namespace` request returns 200

![image](https://github.com/eclipse/che/assets/83611742/608c8bfb-00a8-4765-8192-a073900060e9)

The `Gap-Auth` and `Gap-Upstream-Address` headers seems to be set by `oauth-proxy` container in the che-gateway pod:
https://github.com/openshift/oauth-proxy/blob/55e0cd172625b3419c698014fa233e54021af9a3/oauthproxy.go#L99
https://github.com/openshift/oauth-proxy/blob/55e0cd172625b3419c698014fa233e54021af9a3/oauthproxy.go#L830-L834

</details>

<details><summary>401 error</summary>

![image](https://github.com/eclipse/che/assets/83611742/901a22fc-b55e-4aea-a1c4-67787303f570)

![image](https://github.com/eclipse/che/assets/83611742/7f57a20a-393a-4e84-894d-dbc9ae888734)

In this case, the `namespace` request succeeded, but for whatever reason, the `events` and `pods` request failed with `401`. Which is strange because the `devworkspace` request succeeded.

Just like the `403` case, the `Gap-Auth` and `Gap-Upstream-Address` headers are missing in the response header.

</details>

**Comment by dkwon17** [comment]
For the 401/403 error that sometimes appears when refreshing the dashboard, it seems to happen on these requests made by the dashboard:

| Request                 | Fulfilled By          |
|-------------------------|-----------------------|
| `namespace/provision`   | Che Server            |
| `namespace`             | Che Server            |
| `events`                | Che Dashboard backend |
| `devworkspacetemplates` | Che Dashboard backend |
| `devworkspaces`         | Che Dashboard backend |
| `pods`                  | Che Dashboard backend |

For this specific example of the error:
<img width="1093" alt="image" src="https://github.com/eclipse/che/assets/83611742/6dee98f0-8c0d-4b02-be3e-8940c5859cf7">


On the screenshot we see the following error code from the response:
```
FST_UNAUTHORIZED
```

which is coming from the dashboard backend [here](https://github.com/eclipse-che/che-dashboard/blob/6b45b26a93848f2cd9a4012492b60162898effb2/packages/dashboard-backend/src/routes/api/helpers/getToken.ts#L21), meaning that the authorization header which contains the bearer token is missing.

It's strange that in the screenshot, we see that requests to fetch `devworkspacetemplates` , `devworkspaces` , and `pods` succeeded, but only fetching `events` failed.

For the `events` request to reach the dashboard backend api (and same with the che server api), the request needs to go through the `oauth-proxy` container ([which sets the bearer token in the request header](https://github.com/openshift/oauth-proxy/blob/55e0cd172625b3419c698014fa233e54021af9a3/oauthproxy.go#L828)), and then go through the `gateway` (traefik) container ([which uses a custom middleware to set the authorization header with the token set by](https://github.com/eclipse-che/che-operator/blob/6cf8df2b860b81ea53b3bc98aafe6c25a313aa7c/pkg/deploy/dashboard/dashboard.go#L129) `oauth-proxy`) within the `che-gateway` pod.

Because of `oauth-proxy` and `gateway`, the bearer token should be available in the request's header.

But since the request to the dashboard backend ultimately reaches [here](https://github.com/eclipse-che/che-dashboard/blob/6b45b26a93848f2cd9a4012492b60162898effb2/packages/dashboard-backend/src/routes/api/helpers/getToken.ts#L21), it can either mean:
* oauth-proxy container did not set the access token in the request header
* custom traefik middleware was unable to set the authorization header with the access token set by oauth-proxy container
* somehow the authorization header is missing when the request reaches dashboard-backend 
* maybe something else

**Comment by tolusha** [comment]
I am wondering if we can test latest images [1] [2], maybe it will fix our problem.
[1] https://quay.io/openshift/origin-oauth-proxy
[2] https://quay.io/openshift/origin-kube-rbac-proxy

**Comment by ibuziuk** [comment]
@dkwon17 @tolusha we should probably investigate how ArgoCD manages the OAuth. The UX flow is very similar to what we do with Eclipse Che

**Comment by Mbd06b** [comment]
I was personally encountering this _unauthorized_ issue because I had _**multiple**_ nodes in my cluster, and not all the nodes had my ODIC issuer arguments applied. 

I edited the API Server Configuration in the configuration file for my Kubernetes API server (under /var/snap/microk8s/current/args/kube-apiserver for microk8s).

OIDC Options: Include the following lines (adjusting for your specific Keycloak configuration):

--oidc-issuer-url=https://<KEYCLOAK_DOMAIN>/auth/realms/<REALM_NAME>
--oidc-client-id=<CLIENT_ID>
--oidc-username-claim=email
--oidc-groups-claim=groups 

While I ran this from the node I built the cluster from, I discovered that **I needed to apply these --odic args on each node,** and restart the node for the new configuration to apply, and then my auth issues with login and the devworkspace were resolved.
.
`sudo microk8s stop && sudo microk8s start`

Just for your consideration if there's a similar multi-node situation to consider.
